pkg/collection/entry/metadata.go

```diff
@@ -0,0 +1,50 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package entry
+
+import (
+	"bytes"
+	"fmt"
+)
+
+// Metadata provides mime type and filename to file entry.
+type Metadata struct {
+	MimeType string
+	Filename string
+}
+
+// NewMetadata creates a new Metadata.
+func NewMetadata(fileName string) *Metadata {
+	return &Metadata{
+		Filename: fileName,
+	}
+}
+
+// WithMimeType adds mime type to Metadata.
+func (m *Metadata) WithMimeType(mimeType string) *Metadata {
+	m.MimeType = mimeType
+	return m
+}
+
+// MarshalBinary implements encoding.BinaryMarshaler
+func (m *Metadata) MarshalBinary() ([]byte, error) {
+	b := []byte(m.Filename)
+	if m.MimeType != "" {
+		bs := append(b, []byte{0x00}...)
```

Comment:
aren't strings terminated by a `nil` character? wouldn't this be a problem?

Reply:
This encoding is very arbitrary even if it is simple. It means that we need to escape 0 character in the filename and mime type. My opinion is that json encoding provides more benefits as it is well known and we will have easier time in the future when we extend metadata and want to ensure backward compatibility.

Reply:
You may be right, I will double check if any text encoding would use zero byte. We would want `utf-8` filenames I guess. `0x00` as delimiter of string is a pretty normal thing, though.

Reply:
> 0x00 as delimiter of string is a pretty normal thing, though.

so then when you split, you get an extra entry no? since this byte constitutes an empty string then?

Reply:
I'll just fastfood json it and end of discussion ;)

Reply:
> It means that we need to escape 0 character in the filename and mime type.

that being said; this is not "0" character, this is 0 byte. There are no zero-bytes in string encoding, except to define end-of-string.

Reply:
Yes, of course, I did not mean "0" character, but character with all 0 bytes, just as you wrote 0x00, which is the same as 0.